editor/src/components/canvas/direct-dom-lookups.ts

```diff
@@ -0,0 +1,28 @@
+import { UTOPIA_PATH_KEY } from '../../core/model/utopia-constants'
+import { CanvasContainerID } from './canvas-types'
+import { getDeepestPathOnDomElement } from '../../core/shared/uid-utils'
+import * as EP from '../../core/shared/element-path'
+import type { ElementPath } from 'utopia-shared/src/types'
+
+export type FromElement<T> = (element: HTMLElement) => T
+
+export function getFromElement<T>(path: ElementPath, fromElement: FromElement<T>): T | undefined {
+  const pathString = EP.toString(path)
+  const elements = document.querySelectorAll(
+    `#${CanvasContainerID} [${UTOPIA_PATH_KEY}^="${pathString}"]`,
```

Comment:
in a nested query it's slightly more efficient to use 
```js
document.getElementById(CanvasContainerID).querySelectorAll(`[${UTOPIA_PATH_KEY}^="${pathString}"]`)
```
(since there is no upwards lookup),
but it's really minor.